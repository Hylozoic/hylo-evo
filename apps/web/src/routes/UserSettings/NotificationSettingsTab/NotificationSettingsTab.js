import React, { Component } from 'react'
import { useTranslation, withTranslation } from 'react-i18next'
import cx from 'classnames'
import { compact } from 'lodash/fp'
import PropTypes from 'prop-types'
import Tooltip from 'components/Tooltip'
import Icon from 'components/Icon'
import Loading from 'components/Loading'
import Select from 'components/Select'
import { bgImageStyle } from 'util/index'

import classes from './NotificationSettingsTab.module.scss'

const allGroupsLogo = '/hylo-merkaba.png'

const { object, func } = PropTypes

const iOSAppURL = 'https://itunes.apple.com/app/appName/id1002185140'
const androidAppURL = 'https://play.google.com/store/apps/details?id=com.hylo.hyloandroid'

class NotificationSettingsTab extends Component {
  static propTypes = {
    currentUser: object,
    updateUserSettings: func
  }

  updateMessageSettings = changes => {
    const { messageSettings, updateUserSettings } = this.props
    const newMessageSettings = {
      ...messageSettings,
      ...changes
    }
    let dmNotifications
    if (newMessageSettings['sendEmail'] && newMessageSettings['sendPushNotifications']) {
      dmNotifications = 'both'
    } else if (newMessageSettings['sendEmail']) {
      dmNotifications = 'email'
    } else if (newMessageSettings['sendPushNotifications']) {
      dmNotifications = 'push'
    } else {
      dmNotifications = 'none'
    }
    updateUserSettings({
      settings: {
        dmNotifications
      }
    })
  }

  updateAllGroups = changes => {
    const { memberships, updateAllMemberships } = this.props
    updateAllMemberships(memberships.map(m => m.group.id), changes)
  }

  updateAllGroupsAlert = changes => {
    const { t } = this.props

    const key = ('sendEmail' in changes) ? 'sendEmail' : 'sendPushNotifications'

    const type = key === 'sendEmail' ? t('Email') : t('Push Notifications')
    const onOrOff = changes[key] ? t('ON') : t('OFF')
    const numGroups = this.props.memberships.length

    if (window.confirm(t('You wish to turn {{onOrOff}} {{type}} for all groups? This will affect {{numGroups}} {{groups}}.', { onOrOff, type, numGroups, groups: numGroups === 1 ? t('group') : t('groups') }))) {
      this.updateAllGroups(changes)
    }
  }

  render () {
    const {
      currentUser, updateUserSettings, memberships, updateMembershipSettings, allGroupsSettings,
      messageSettings, t
    } = this.props

    if (!currentUser) return <Loading />

    const { settings = {}, hasDevice } = currentUser

    const updateSetting = setting => value => {
      updateUserSettings({ settings: { [setting]: value } })
    }

    const notificationOptions = compact([
      { id: 'none', label: t('None') },
      { id: 'email', label: t('Email') },
      hasDevice && { id: 'push', label: t('Mobile App') },
      hasDevice && { id: 'both', label: t('Both') }
    ])

    const getSetting = setting => {
      if (!hasDevice && settings[setting] === 'both') return 'email'
      return settings[setting]
    }
    return (
      <div>
        <div className={classes.title}><Icon name='Notifications' />{t('Notifications')}</div>
        <div className={classes.globalSetting}>
          <div className={classes.prompt}>{t('How often would you like to receive email digests for new posts in your groups and saved searches?')}
          </div>
          <div className={classes.settingSelect}>
            <div className={classes.selectExplanation}>{t('Send me a digest')}</div>
            <Select
              onChange={updateSetting('digestFrequency')}
              selected={settings.digestFrequency}
              options={[
                { id: 'daily', label: t('Daily') },
                { id: 'weekly', label: t('Weekly') },
                { id: 'never', label: t('Never') }
              ]} />
          </div>
        </div>
        <div className={classes.globalSetting}>
          <div className={classes.prompt}>{t('Would you like to receive a notification for each new post in your groups?')}
          </div>
          <div className={classes.settingSelect}>
            <Select
              onChange={updateSetting('postNotifications')}
              selected={settings.postNotifications}
              options={[
                { id: 'none', label: t('No Posts') },
                { id: 'important', label: t('Important Posts (Announcements & Mentions)') },
                { id: 'all', label: t('Every Post') }
              ]} />
          </div>
        </div>
        <div className={classes.globalSetting}>
          <div className={classes.prompt}>
            {t('How would you like to receive notifications about')}&nbsp;
            {t('new comments on posts you\'re following?')}
          </div>
          <div className={classes.settingSelect}>
            <div className={classes.selectExplanation}>{t('Notify me via')}</div>

            <Select
              onChange={updateSetting('commentNotifications')}
              selected={getSetting('commentNotifications')}
              options={notificationOptions} />
          </div>
        </div>
        <div>
          <div className={classes.individualGroups}>{t('NOTIFICATIONS')}</div>

          <MessageSettingsRow
            settings={messageSettings}
            updateMessageSettings={this.updateMessageSettings} />

          <div className={classes.individualGroups}>{t('GROUP NOTIFICATIONS')}</div>

          <AllGroupsSettingsRow
            settings={allGroupsSettings}
            updateAllGroups={this.updateAllGroupsAlert} />
          {memberships.map(membership => <MembershipSettingsRow
            key={membership.id}
            membership={membership}
            updateMembershipSettings={changes => updateMembershipSettings(membership.group.id, changes)} />)}
        </div>

        <div className={classes.help}>
          <p className={classes.helpParagraph}>
            {t('Download our')}{' '}<a href={iOSAppURL} rel='noreferrer' target='_blank'>iOS</a>
            {' '}{t('or')}{' '}
            <a href={androidAppURL} rel='noreferrer' target='_blank'>Android</a>
            {t(' app to receive push notifications.')}
          </p>
        </div>

        <Tooltip
          delay={250}
          id='helpTip'
          position='top'
        />
      </div>
    )
  }
}

export function MessageSettingsRow ({ settings, updateMessageSettings }) {
  const { t } = useTranslation()
  return (
    <SettingsRow
      iconName='Messages'
      name={t('Messages')}
      settings={settings}
      update={updateMessageSettings}
    />
  )
}

export function AllGroupsSettingsRow ({ settings, updateAllGroups }) {
  const { t } = useTranslation()
  return (
    <SettingsRow
      imageUrl={allGroupsLogo}
      name={t('All Groups')}
      settings={settings}
      update={updateAllGroups}
    />
  )
}

export function MembershipSettingsRow ({ membership, updateMembershipSettings }) {
  return (
    <SettingsRow
      imageUrl={membership.group.avatarUrl}
      name={membership.group.name}
      settings={membership.settings}
      update={updateMembershipSettings}
    />
  )
}

export class SettingsRow extends React.Component {
  render () {
    const { iconName, imageUrl, name, settings, update } = this.props

    const imageStyle = bgImageStyle(imageUrl)

    return (
      <div className={classes.settingsRow}>
        <div className={classes.nameRow}>
          {iconName && <Icon name={iconName} className={classes.avatarIcon} />}
          {!iconName && <div className={classes.groupAvatar} style={imageStyle} />}
          <span className={classes.name}>{name}</span>
        </div>
        <div className={classes.iconRow}>
          <SettingsIcon settingKey='sendPushNotifications' name='PushNotification' settings={settings} update={update} />
          <SettingsIcon settingKey='sendEmail' name='EmailNotification' settings={settings} update={update} />
        </div>
      </div>
    )
  }
}

export function SettingsIcon ({ settingKey, name, update, settings }) {
  const settingStatus = settings[settingKey] ? 'On' : 'Off'
  const { t } = useTranslation()

  return (
    <div
      className={cx(classes.settingControls, { [classes.highlightIcon]: settings[settingKey] })}
      onClick={() => update({ [settingKey]: !settings[settingKey] })}
      data-tooltip-content={`Turn ${name === 'EmailNotification' ? 'Email' : 'Mobile Push'} Notifications ${settings[settingKey] ? t('Off') : t('On')}`}
      data-tooltip-id='helpTip'
    >
      <Icon name={name} className={cx(classes.icon, { [classes.highlightIcon]: settings[settingKey] })} />
      <span className={classes.settingStatus}>{t(settingStatus)}</span>
    </div>
  )
}
export default withTranslation()(NotificationSettingsTab)
