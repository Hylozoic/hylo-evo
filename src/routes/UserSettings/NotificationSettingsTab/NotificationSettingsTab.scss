.title {
  height: 60px;
  line-height: 60px;
  text-align: center;
  border-radius: 4px;
  margin-bottom: 25px;
  composes: hdr-display from 'css/typography.scss';
  background-color: $color-catskill-white;
  border: none;
  width: 100%;
}

.prompt {
  composes: bdy-drk-sm from 'css/typography.scss';
  margin-bottom: 20px;
}

.help {
  padding-top: 40px;
  composes: bdy-lt-sm from 'css/typography.scss';
  a {
    color: $color-rhino;
  }
  border-top: $light-border;
}

.help-paragraph {
  margin-bottom: 20px;
}

///// probably trash

.banner {
  width: 100%;
  height: 165px;
  border-radius: 4px;
  background-position: center;
  background-size: cover;
  padding-top: 5px;
  padding-left: 5px;
}

.change-banner-button {
  /* empty */
}

.avatar {
  display: flex;
  align-items: flex-end;
  width: 140px;
  height: 140px;
  border-radius: 50%;
  background-position: center;
  background-size: cover;
  margin: -70px auto $space-8x auto;
}

.change-avatar-button {
  margin: 0 auto;
  width: 30px;
}

.control-label {
  display: block;
  composes: tag from 'css/typography.scss';
  color: $color-rhino-30;
  margin-bottom: $space-1x;
}

.social-label {
  composes: control-label;
  margin-top: $space-9x;
}

.button-row {
  display: flex;
  width: 100%;
  margin-top: $space-9x;
}

.save-button {
  margin: 0 auto;
}

// Control
.control {
  margin-bottom: $space-4x;
}

.control-input {
  width: 100%;
  border: none;
  border-bottom: 1px solid $color-ghost;
  composes: bdy-drk-sm from 'css/typography.scss';
  font-size: 16px;
  padding-bottom: $space-2x;
  resize: none;
  &:focus {
    outline: none;
  }
}

textarea.control-input {
  resize: vertical;
  height: $space-20x;
}

// SocialControl
.linked {
  border-bottom: 1px solid $color-caribbean-green;
}

.social-control-label {
  composes: control-input;
  display: flex;
}

.link-button {
  margin-left: auto;
  composes: text-button from 'css/typography.scss';
}
