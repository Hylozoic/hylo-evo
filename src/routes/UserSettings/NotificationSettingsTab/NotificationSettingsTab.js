import React, { PropTypes, Component } from 'react'
import './NotificationSettingsTab.scss'
import Icon from 'components/Icon'
import Loading from 'components/Loading'
import { Link } from 'react-router-dom'
import { compact } from 'lodash/fp'
const { object, func } = PropTypes

const iOSAppURL = 'https://itunes.apple.com/app/appName/id1002185140'
const androidAppURL = 'https://play.google.com/store/apps/details?id=com.hylo.hyloandroid'

export default class NotificationSettingsTab extends Component {
  static propTypes = {
    currentUser: object,
    updateUserSettings: func
  }

  render () {
    const { currentUser, updateUserSettings } = this.props

    if (!currentUser) return <Loading />

    const { settings, hasDevice } = currentUser

    const updateSetting = setting => value => {
      updateUserSettings({settings: {[setting]: value}})
    }

    var notificationOptions = compact([
      {id: 'none', label: 'None'},
      {id: 'email', label: 'Email'},
      hasDevice && {id: 'push', label: 'Mobile App'},
      hasDevice && {id: 'both', label: 'Both'}
    ])

    return <div>
      <div styleName='title'>Notifications</div>
      <div styleName='prompt'>How often would you like to receive email digests
        for new posts in your communities?</div>
      <Select
        onChange={updateSetting('digestFrequency')}
        selected={settings['digestFrequency']}
        options={[
          {id: 'daily', label: 'Daily'},
          {id: 'weekly', label: 'Weekly'},
          {id: 'never', label: 'Never'}
        ]} />
      <div styleName='prompt'>How would you like to receive notifications about
        new direct messages?</div>
      <Select
        onChange={updateSetting('dmNotifications')}
        selected={settings['dmNotifications']}
        options={notificationOptions} />
      <div styleName='prompt'>How would you like to receive notifications about
      new comments on posts you're following?</div>
      <Select
        onChange={updateSetting('commentNotifications')}
        selected={settings['commentNotifications']}
        options={notificationOptions} />
      <div styleName='help'>
        <p styleName='help-paragraph'>
          Download our <a href={iOSAppURL} target='_blank'>iOS</a>&nbsp;
          or <a href={androidAppURL} target='_blank'>Android</a> app to
          receive push notifications.
        </p>
        <p styleName='help-paragraph'>
          See the <Link to='/settings/communities' replace>Communities</Link> section
          to change notifications for an individual community.
        </p>
      </div>
    </div>
  }
}

export function Select ({ options, selected, onChange }) {
  return <span styleName='select-wrapper'>
    <select styleName='select' onChange={e => onChange(e.target.value)} value={selected}>
      {options.map(({id, label}) => <option value={id} key={id}>
        {label}
      </option>)}
    </select>
    <Icon name='ArrowDown' styleName='icon' />
  </span>
}
