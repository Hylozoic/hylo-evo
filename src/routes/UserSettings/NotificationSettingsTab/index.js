import NotificationSettingsTab from './NotificationSettingsTab'

export default NotificationSettingsTab
