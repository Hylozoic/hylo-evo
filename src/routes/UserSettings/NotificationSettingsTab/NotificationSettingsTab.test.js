import NotificationSettingsTab from './NotificationSettingsTab'
import { shallow } from 'enzyme'
import React from 'react'

describe('NotificationSettingsTab', () => {
  const currentUser = {
    hasDevice: true,
    settings: {
      digestFrequency: 'daily',
      dmNotifications: 'none',
      commentNotifications: 'email'
    }
  }

  it('renders correctly', () => {
    const wrapper = shallow(<NotificationSettingsTab
      currentUser={currentUser}
      updateUserSettings={() => {}} />)
    expect(wrapper).toMatchSnapshot()
  })

  it("hides mobile options if user doesn't have device", () => {
    const wrapper = shallow(<NotificationSettingsTab currentUser={{
      ...currentUser,
      hasDevice: false
    }} />)
    expect(wrapper.find('Select').at(1).prop('options')).toHaveLength(2)
  })

  it("does the right thing with 'both' if user doesn't have device", () => {
    const wrapper = shallow(<NotificationSettingsTab currentUser={{
      ...currentUser,
      settings: {
        ...currentUser.settings,
        dmNotifications: 'both'
      },
      hasDevice: false
    }} />)

    expect(wrapper.find('Select').at(1).prop('selected')).toEqual('email')
  })
})
