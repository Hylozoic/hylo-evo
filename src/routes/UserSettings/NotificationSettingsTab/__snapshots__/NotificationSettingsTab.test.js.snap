// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllCommunitiesSettingsRow renders correctly 1`] = `
<SettingsRow
  imageUrl="/hylo-merkaba.png"
  name="All Communities"
  settings={
    Object {
      "sendEmail": true,
    }
  }
  update={[Function]}
/>
`;

exports[`MembershipSettingsRow renders correctly 1`] = `
<SettingsRow
  imageUrl="foo.png"
  name="Foomunity"
  settings={
    Object {
      "sendEmail": true,
    }
  }
  update={[Function]}
/>
`;

exports[`MessageSettingsRow renders correctly 1`] = `
<SettingsRow
  iconName="Messages"
  name="Messages"
  settings={
    Object {
      "sendEmail": true,
    }
  }
  update={[Function]}
/>
`;

exports[`NotificationSettingsTab renders correctly 1`] = `
<div>
  <div
    data-stylename="title"
  >
    Notifications
  </div>
  <div
    data-stylename="prompt"
  >
    How often would you like to receive email digests for new posts in your communities?
  </div>
  <Select
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": "daily",
          "label": "Daily",
        },
        Object {
          "id": "weekly",
          "label": "Weekly",
        },
        Object {
          "id": "never",
          "label": "Never",
        },
      ]
    }
    selected="daily"
  />
  <div
    data-stylename="prompt"
  >
    How would you like to receive notifications about new comments on posts you're following?
  </div>
  <Select
    onChange={[Function]}
    options={
      Array [
        Object {
          "id": "none",
          "label": "None",
        },
        Object {
          "id": "email",
          "label": "Email",
        },
        Object {
          "id": "push",
          "label": "Mobile App",
        },
        Object {
          "id": "both",
          "label": "Both",
        },
      ]
    }
    selected="email"
  />
  <div>
    <MessageSettingsRow
      updateMessageSettings={[Function]}
    />
    <AllCommunitiesSettingsRow
      updateAllCommunities={[Function]}
    />
    <MembershipSettingsRow
      key="1"
      membership={
        Object {
          "id": 1,
        }
      }
      updateMembershipSettings={[Function]}
    />
    <MembershipSettingsRow
      key="2"
      membership={
        Object {
          "id": 2,
        }
      }
      updateMembershipSettings={[Function]}
    />
  </div>
  <div
    data-stylename="help"
  >
    <p
      data-stylename="help-paragraph"
    >
      Download our 
      <a
        href="https://itunes.apple.com/app/appName/id1002185140"
        target="_blank"
      >
        iOS
      </a>
      Â  or 
      <a
        href="https://play.google.com/store/apps/details?id=com.hylo.hyloandroid"
        target="_blank"
      >
        Android
      </a>
       app to receive push notifications.
    </p>
  </div>
</div>
`;

exports[`SettingsIcon renders correctly 1`] = `
<Icon
  data-stylename="icon highlightIcon"
  name="EmailNotification"
  onClick={[Function]}
/>
`;

exports[`SettingsRow renders correctly 1`] = `
<div
  data-stylename="settingsRow"
>
  <div
    data-stylename="nameRow"
  >
    <div
      data-stylename="communityAvatar"
      style={Object {}}
    />
    <span
      data-stylename="name"
    />
    <Icon
      data-stylename="arrowIcon"
      name="ArrowDown"
      onClick={[Function]}
    />
  </div>
</div>
`;
